Libs/Optimize/ParticleSystem/VtkMeshWrapper.cpp

```diff
@@ -0,0 +1,822 @@
+{
+  double point[3];
+  point[0] = pointa[0];
+  point[1] = pointa[1];
+  point[2] = pointa[2];
+  double closest_point[3];
+  this->GetTriangleForPoint(point, idx, closest_point);
+
+  VtkMeshWrapper::PointType out;
+  out[0] = closest_point[0];
+  out[1] = closest_point[1];
+  out[2] = closest_point[2];
+  return out;
+
+/*
+
+  double point[3];
+  point[0] = pointa[0];
+  point[1] = pointa[1];
+  point[2] = pointa[2];
+
+  double closest_point[3]; //the coordinates of the closest point will be returned here
+  double closest_point_dist2; //the squared distance to the closest point will be returned here
+  vtkIdType cell_id; //the cell id of the cell containing the closest point will be returned here
+  int sub_id; //this is rarely used (in triangle strips only, I believe)
+
+  this->cell_locator_->FindClosestPoint(point, closest_point, cell_id, sub_id, closest_point_dist2);
+
+  VtkMeshWrapper::PointType out;
+  out[0] = closest_point[0];
+  out[1] = closest_point[1];
+  out[2] = closest_point[2];
+
+  // update cache
+  if (idx > 0) {
+    if (idx >= particle2tri_.size()) {
+      particle2tri_.resize(idx + 1, 0);
+    }
+
+    particle2tri_[idx] = cell_id;
+  }
+
+  return out;
+  */
+}
+
+//---------------------------------------------------------------------------
+VtkMeshWrapper::PointType VtkMeshWrapper::GetPointOnMesh() const
+{
+  int face_index = 0;
+
+  double* point = this->poly_data_->GetPoint(face_index);
+
+  PointType p;
+  p[0] = point[0];
+  p[1] = point[1];
+  p[2] = point[2];
+
+  return p;
+}
+
+//---------------------------------------------------------------------------
+int VtkMeshWrapper::GetTriangleForPoint(const double pt[3], int idx, double closest_point[3]) const
+{
+  // given a guess, just check whether it is still valid.
+  if (idx >= 0) {
+    // ensure that the cache has enough elements. this will never be resized to more than the number of particles,
+    if (idx >= particle2tri_.size()) {
+      particle2tri_.resize(idx + 1, -1);
+    }
+
+    const int guess = particle2tri_[idx];
+
+
+    if (guess != -1 && this->IsInTriangle(pt, guess)) {
+      closest_point[0] = pt[0];
+      closest_point[1] = pt[1];
+      closest_point[2] = pt[2];
+      return guess;
+    }
+
+
+/*
+    if (guess > 0) {
+      closest_point[0] = pt[0];
+      closest_point[1] = pt[1];
+      closest_point[2] = pt[2];
+      return guess;
+    }
+*/
+  }
+
+  //double closest_point[3];//the coordinates of the closest point will be returned here
+  double closest_point_dist2; //the squared distance to the closest point will be returned here
+  vtkIdType cell_id; //the cell id of the cell containing the closest point will be returned here
+  int sub_id; //this is rarely used (in triangle strips only, I believe)
+
+  this->cell_locator_->FindClosestPoint(pt, closest_point, cell_id, sub_id, closest_point_dist2);
+
+  if (idx >= 0) {
+    particle2tri_[idx] = cell_id;
```

Comment:
we could end up in a situation where particle2tri is valid, but particle2normal is invalid. does this manifest as a problem, or does the normal get updated before being used?

Reply:
You are right, adding another check based on position